### PR TITLE
Cluster Config Deserialisation Tests

### PR DESCRIPTION
**Pull request by jeremyk-91**
**Goals (and why)**:
- Add full deserialisation tests for polymorphic deserialisation of `ClusterConfiguration`
- Gain more confidence that `ClusterConfiguration` deserialisation works properly

**Implementation Description (bullets)**:
- Add several yaml blobs which contain possible cluster configurations, including one without type information (because this is consistent with the legacy format)

**Concerns (what feedback would you like?)**:
- One of the tests exposes #2821 but I don't see a simple way to fix it so I left it as it is.
- The K8s test here is not ideal, not sure if there's a better way.

**Where should we start reviewing?**:
- It's probably easiest to look at each of the small yaml blobs.

**Priority (whenever / two weeks / yesterday)**: whenever; I did a manual test on the PR #2794 and am pretty confident it works.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2822)

### CODE REVIEW

File: timelock-agent/src/test/java/com/palantir/timelock/config/ClusterConfigurationDeserializationTest.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+    @Test
+    public void throwsOnMalformedClusterConfiguration() throws IOException {
+        assertThatThrownBy(() -> deserializeClusterConfiguration(CLUSTER_CONFIG_MALFORMED))
+                .isInstanceOf(UnrecognizedPropertyException.class)
+                .hasMessageContaining("my-node-set");
+    }
+
+    @Test
+    public void doesNotDeserializeKubernetesClusterConfigurationAsDefaultClusterConfiguration() throws IOException {
```
Comment (gmaretic) [inline timelock-agent/src/test/java/com/palantir/timelock/config/ClusterConfigurationDeserializationTest.java]: Do we already have a test with a k8s config that deserializes correctly?

File: timelock-agent/src/test/java/com/palantir/timelock/config/ClusterConfigurationDeserializationTest.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        assertDefaultClusterConfigurationCorrect(deserializeClusterConfiguration(CLUSTER_CONFIG_NO_TYPE_INFO));
+    }
+
+    @Test
+    public void canDeserializeClusterConfigurationWithDefaultTypeInformation() throws IOException {
+        assertDefaultClusterConfigurationCorrect(deserializeClusterConfiguration(CLUSTER_CONFIG_DEFAULT_TYPE_INFO));
+    }
+
+    @Test
+    public void canDeserializeClusterConfigurationWithInvalidTypeInformation() throws IOException {
```
Comment (gmaretic) [inline timelock-agent/src/test/java/com/palantir/timelock/config/ClusterConfigurationDeserializationTest.java]: I'm a bit uneasy merging a test that tests that something is wrong :)

Maybe rather write the test you want to have and then @Ignore it?

Comment (jeremyk-91) [reply timelock-agent/src/test/java/com/palantir/timelock/config/ClusterConfigurationDeserializationTest.java]: Ah, okay. Yeah, the ignore-test makes sense.

File: timelock-agent/src/test/java/com/palantir/timelock/config/ClusterConfigurationDeserializationTest.java (hunk 1 above)
Comment (jeremyk-91) [reply timelock-agent/src/test/java/com/palantir/timelock/config/ClusterConfigurationDeserializationTest.java]: There are tests that the already-deserialized k8s config is right. As mentioned in the comment, unfortunately we rely on a static instance that finds the local hostname and uses it. In terms of deserialisation unfortunately not 😞

**Pull request merged**